### PR TITLE
Support metric label selector for custom metrics

### PR DESCRIPTION
**Pull request by astefanutti**
This PR requires the following to be merged for it to compile and before being merged:
- ~~kubernetes/kubernetes#69806~~
- #34

### CODE REVIEW

File: pkg/apiserver/installer/cmhandlers.go [hunk 1]
```diff
@@ -64,6 +64,20 @@ func (ch *CMHandlers) registerResourceHandlers(a *MetricsAPIInstaller, ws *restf
 		return err
 	}
 
+	listOptions, _, _ := lister.NewListOptions()
```
Comment (DirectXMan12) [inline pkg/apiserver/installer/cmhandlers.go]: please add a comment clarifying why we have both list options and extra list options

Comment (astefanutti) [reply pkg/apiserver/installer/cmhandlers.go]: As the standard `ListOptions` is defined as a struct and not an interface, I don't know of a way to use a single object that could be extended with polymorphism. So there is a standard list options as defined by the `ListOptions` struct, plus the extra options defined by the CM adapter.

I'll add a comment in the code when the question where the `ListerWithOptions` should go is resolved.

**Comment by DirectXMan12** [comment]
Yep, didn't mean to close it, sorry about that

**Comment by DirectXMan12** [comment]
> Thanks for the review. Let me know how we can move the question of the `ListerWithOptions` forward

I suspect moving the changes into the CM adapter codebase might be the best option, if it's not too hairy.  Alternatively, show up at one of the SIG APIMachinery meetings in person and ask, perhaps?  That's generally a quick way to get a resolution.  If you have concrete evidence of the maintenance burden/duplicate code, it'd be good evidence for why we should have in in core API machinery.

File: pkg/apiserver/installer/cmhandlers.go (hunk 1 above)
Comment (DirectXMan12) [reply pkg/apiserver/installer/cmhandlers.go]: My initial suggestion would be to say that custom list options should `,inline` embed the base ListOptions, but I think your idea is actually fine as well :-).

**Comment by astefanutti** [comment]
Sounds good. I'll try to show up at the APIMachinery SIG ASAP. In the meantime, I'll upgrade this PR by integrating the required bits and we'll have material to iterate on.

**Comment by astefanutti** [comment]
I've updated the PR with 3ec1933c4dec404865066ae3782620db060fef8a that contains the bits required from the API server, so that removes the pre-requisite on kubernetes/kubernetes#69806.

**Comment by metalmatze** [comment]
Looking at the code this lgtm. I'm missing a higher level description of what actually is going to change with this PR to be honest.

**Comment by brancz** [comment]
While I see the use for this, I'm wondering if it's actually something we should pursue. It maps somewhat ok to the Prometheus data model, but in Prometheus different labels means a different time-series, so should a list return multiple results with the same metrics name but different labels? My feeling is that the changes we have been thinking about regarding configuration to define what a custom metric is to be converted into in Prometheus might be more pragmatic (in regards to the prometheus-k8s-adapter).

**Comment by DirectXMan12** [comment]
drive-by comment: the adapter *probably* should try and support most/all of the features in the metrics APIs

**Comment by kawych** [comment]
@astefanutti, can we proceed with this PR? I can approve if we get confirmation from @brancz that it's OK to pursue.

@brancz @DirectXMan12 
> drive-by comment: the adapter probably should try and support most/all of the features in the metrics APIs

100% agreed. It's really great that this adapter is refactored so that it can be used as the library, but especially now it will be super useful to support all the features, instead of all adapters implementing them for themselves.

I see a discussion on kubernetes/kubernetes#69806 about supporting custom options - it would be ideal if apiserver package could support our use case. I think we should return to the discussion to understand what are the issues with that - if these are serious concerns, maybe this proposal should be reconsidered/updated: https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/0032-enhance-hpa-metrics-specificity.md#alternatives

**Comment by brancz** [comment]
cc @s-urbaniak (he has been working more closely on metrics API things lately than me so I’d like to hear his opinion).

**Comment by astefanutti** [comment]
@kawych happy to update and proceed with the PR, if we can get an agreement on how to deal with custom options.

File: pkg/apiserver/endpoints/handlers/get.go [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	"k8s.io/apiserver/pkg/endpoints/handlers"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	utiltrace "k8s.io/apiserver/pkg/util/trace"
+
+	cm_rest "github.com/kubernetes-incubator/custom-metrics-apiserver/pkg/apiserver/registry/rest"
+)
+
+func ListResourceWithOptions(r cm_rest.ListerWithOptions, rw rest.Watcher, scope handlers.RequestScope, forceWatch bool, minRequestTimeout time.Duration) http.HandlerFunc {
```
Comment (kawych) [inline pkg/apiserver/endpoints/handlers/get.go]: AFAIU the watch is not supported and forceWatch is unused, please remove

File: pkg/apiserver/registry/rest/rest.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	// This object must be a pointer type for use with Codec.DecodeInto([]byte, runtime.Object)
+	NewList() runtime.Object
+
+	// List selects resources in the storage which match to the selector. 'options' can be nil.
+	// The extraOptions object passed to it is of the same type returned by the NewListOptions
+	// method.
+	List(ctx context.Context, options *metainternalversion.ListOptions, extraOptions runtime.Object) (runtime.Object, error)
+
+	// NewListOptions returns an empty options object that will be used to pass extra options
+	// to the List method. It may return a bool and a string, if true, the
```
Comment (kawych) [inline pkg/apiserver/registry/rest/rest.go]: Can you please explain better the third returned value? AFAIU the second one indicates whether the third one is present?

File: pkg/apiserver/endpoints/handlers/get.go [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+				}
+			} else {
+				opts.FieldSelector = nameSelector
+			}
+		}
+
+		// Log only long List requests (ignore Watch).
+		defer trace.LogIfLong(500 * time.Millisecond)
+		trace.Step("About to List from storage")
+		extraOpts, subpath, subpathKey := r.NewListOptions()
```
Comment (kawych) [inline pkg/apiserver/endpoints/handlers/get.go]: nit: subpath is not the best name for boolean value, maybe hasSubpath?

**Comment by astefanutti** [comment]
@kawych thanks for the review. I’ll be AFK until end of this week. I’ll update the PR next week with your comments.

File: pkg/apiserver/installer/installer.go [hunk 5]
```diff
@@ -286,3 +286,9 @@ func restfulListResource(r rest.Lister, rw rest.Watcher, scope handlers.RequestS
 		handlers.ListResource(r, rw, scope, forceWatch, minRequestTimeout)(res.ResponseWriter, req.Request)
```
[the remark this answers is not in the document]
Comment (astefanutti) [reply pkg/apiserver/installer/installer.go]: Yes, for external metrics.

File: pkg/apiserver/endpoints/handlers/get.go (hunk 2 above)
Comment (astefanutti) [reply pkg/apiserver/endpoints/handlers/get.go]: done

File: pkg/apiserver/registry/rest/rest.go (hunk 3 above)
Comment (astefanutti) [reply pkg/apiserver/registry/rest/rest.go]: Your understanding is correct. It is taken verbatim from `NewGetterOptions`.

File: pkg/apiserver/endpoints/handlers/get.go (hunk 4 above)
Comment (astefanutti) [reply pkg/apiserver/endpoints/handlers/get.go]: done

**Comment by astefanutti** [comment]
I've just rebased the PR and updated it with the review.

**Comment by s-urbaniak** [comment]
post-merge lgtm from my side :+1: Sorry, I have been on a long PTO.

**Pull request merged**